packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <unordered_set>
+#include <vector>
+#include <memory>
+#include <jsi/jsi.h>
+#include "PerformanceEntryLinearBuffer.h"
+
+namespace facebook::react {
+
+class PerformanceObserver;
+
+using PerformanceObserverEntryTypeFilter = std::unordered_set<PerformanceEntryType>;
+using PerformanceObserverCallback = std::function<void(PerformanceObserver& observer)>;
+
+/**
+ * Represents subset of spec's `PerformanceObserverInit` that is allowed for multiple types.
+ *
+ * https://w3c.github.io/performance-timeline/#performanceobserverinit-dictionary
+ */
+struct PerformanceObserverObserveMultipleOptions {
+  double durationThreshold = 0.0;
+};
+
+/**
+ * Represents subset of spec's `PerformanceObserverInit` that is allowed for single type.
+ *
+ * https://w3c.github.io/performance-timeline/#performanceobserverinit-dictionary
+ */
+struct PerformanceObserverObserveSingleOptions {
+  bool buffered = false;
+  double durationThreshold = 0.0;
+};
+
+/**
+ * Represents native counterpart of performance timeline PerformanceObserver
+ * class. Each instance has its own entry buffer and can listen for different
+ * performance entry types.
+ *
+ * Entries are pushed to the observer by the `PerformanceEntryReporter` class,
+ * through the `PerformanceObserverRegistry` class which acts as a central hub.
+ *
+ * To create new performance observers, you should use
+ * `PerformanceObserverRegistry::createObserver` which creates the observer,
+ * registers it immediately and unregisters it when it goes out of scope.
+ *
+ * You can still manually create instance of `PerformanceObserver`, but then
+ * you need to manually register and unregister it from registry.
+ */
+class PerformanceObserver: public jsi::NativeState {
+ public:
+  using Ptr = std::shared_ptr<PerformanceObserver>;
+  using WeakPtr = std::weak_ptr<PerformanceObserver>;
```

Comment:
We're kinda moving away from aliases for different classes of pointers, and you might not need them after some of the feedback.

Reply:
Oh, I actually find them useful (looking up `Ptr` alias tells you whenever its shared or unique).